### PR TITLE
swedish ssn cannot have format like XXXXXX-000X

### PR DESCRIPTION
**Pull request by TrueJacobG**
swedish ssn cannot have format like XXXXXX-000X, the end number part has to be between 001 and 999

explanation: https://www4.skatteverket.se/rattsligvagledning/edition/2024.5/330242.html#update_20240802105842

Information obtained with help of @Jakub-Pazio and @Johannestegner

### CODE REVIEW

File: src/test/java/net/datafaker/idnumbers/SwedishIdNumberTest.java [hunk 1]
```diff
@@ -14,6 +14,7 @@ void validSwedishSsn() {
 
     @Test
     void invalidSwedishSsn() {
+        assertThat(SwedenIdNumber.isValidSwedishSsn("020914-0001")).isFalse();
```
Comment (kingthorin) [inline src/test/java/net/datafaker/idnumbers/SwedishIdNumberTest.java:17]: I don't get this, you said 001 to 999 are fine for last 3. So why should 001 be isvalid false?

Or did you mean it should be 3 digits (excluding triple zero) following the hyphen not 4?

Comment (TrueJacobG) [reply src/test/java/net/datafaker/idnumbers/SwedishIdNumberTest.java:17]: I mean last 3 number without checksum digit so 020914-0001 is invalid.

**Review by kingthorin — APPROVED** [review]
Thanks

File: src/main/java/net/datafaker/idnumbers/SwedenIdNumber.java [hunk 2]
```diff
@@ -37,14 +37,23 @@ public String getValidSsn(BaseProviders f) {
...
             + end;
         String idNumber = basePart + calculateChecksum(basePart);
         return new PersonIdNumber(idNumber, birthday, gender(f, request));
     }
 
+    private String generateEndPart(BaseProviders f) {
+        String end;
+        do {
+            end = f.numerify("###");
+        } while (!end.equals("000"));
```
Comment (asolntsev) [inline src/main/java/net/datafaker/idnumbers/SwedenIdNumber.java:52]: Isn't it a bug?
Shouldn't it be the opposite: `while (end.equals("000"));`?

Comment (kingthorin) [reply src/main/java/net/datafaker/idnumbers/SwedenIdNumber.java:52]: Ha, good catch. I think you're right

Comment (TrueJacobG) [reply src/main/java/net/datafaker/idnumbers/SwedenIdNumber.java]: Yeah, you're right 😅. Already fixed

File: src/main/java/net/datafaker/idnumbers/SwedenIdNumber.java [hunk 3]
```diff
@@ -78,6 +87,10 @@ public static boolean isValidSwedishSsn(String ssn) {
             return false;
         }
 
+        if(ssn.startsWith("000", 7)){
```
Comment (kingthorin) [inline src/main/java/net/datafaker/idnumbers/SwedenIdNumber.java:90]:
```suggestion
        if (ssn.startsWith("000", 7)) {
```

**Review by kingthorin — APPROVED** [review]

**Review by asolntsev — CHANGES_REQUESTED** [review]
@TrueJacobG There is a way to generate non-"000" number without a loop:

```java
    private String generateEndPart(BaseProviders f) {
        return "%03d".formatted(f.number().numberBetween(1, 1000));
    }
```

**Review by asolntsev — APPROVED** [review]

**Comment by bodiam** [comment]
Thanks @asolntsev , @kingthorin and of course @TrueJacobG for working on this together, nice improvement!

**Pull request merged**